### DISCUSSION TITLE
Is there a way to check for multiple events (of one tx) in a test case?

### DISCUSSION THREAD

**Discussion by 0xDEnYO** [foundry-rs/foundry#3585] [Help] [answered]
Hey Guys, 

I am wondering if its possible to check one function call for several emitted events, similar to this code here (which I have not managed to get working): 

```
// check for event 1 (which should be emitted by the contract that I am testing)
vm.expectEmit(true, true, true, true);
emit Event1(address parameter 1, uint256 parameter 2);

// check for event 2 (which should be emitted by a contract that is called by the contract that I am testing)
vm.expectEmit(false, false, false, false);
emit Event2(uint256 parameter 1, address parameter 2); 

// call function that should emit both events
actualFunctionCallThatShouldEmitBothEvents();
```

Currently I can only manage to check for one event which limits my ability to check if a function call really does what it should (in my case: 1) emit event in our contract and 2) call another contract that emits an event, too). 
Maybe there are other ways to achieve the same result but I cant think of any.

Also, it seems that when I check for event 2 (the one that is emitted by another contract that my contract calls) it fails all the time....so can I only check for events emitted by the contract I am testing?

Thanks a lot for any advice

**Comment by mds1** [comment]
That should work, I've used `expectEmit` in a similar way before. Can you provide a repo or code to reproduce the issue?

**Comment by 0xDEnYO** [comment]
Hi Matt, 

thanks for your response. 
I was able to solve the issue. I had a problem in the event parameters. Also I was able now to check for events of other contracts.

Another question I was not able to answer yet is:
can I also check if a certain event was NOT emitted (as in hardhat ...to.not.emit(...))?

**Reply by mds1** [reply]
That is being tracked here, but it's not currently supported https://github.com/foundry-rs/foundry/issues/509

**Comment by 0xDEnYO** [comment]
Hey Guys, 

I have a test case where I am checking for two events to be emitted.  The test fails with "Log != expected log" without giving me any details about which parameter or value didnt match, even when run with highest verbose level.

when I comment out either of the two event checks, the test passes. So both events seem fine by themselves. Also the call tree suggests that all event parameters match.

I dont know where else to look for the error. Do you have any idea what could be going wrong here?

Output: 
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/77058885/203321171-3bbde187-2f23-425a-9f77-0c353d5b19d5.png">

Relevant Code: 

![image](https://user-images.githubusercontent.com/77058885/203321260-ade25e11-5e3a-49d0-a0f8-f850a56b1c31.png)

Any idea what could cause this behaviour?





And a suggestion: 
Would it be possible to improve the output of test fails caused by non-matching event parameters or address to something like:

```
                     expected   |   actual    |    CHECK
event argument a          5     |      5      |     ok
event argument b          5     |      6      |     fail
emitting address     0x1234.... |  0x321...   |     fail
```

**Reply by mds1** [reply]
> The test fails with "Log != expected log" without giving me any details about which parameter or value didnt match, even when run with highest verbose level.

Pretty sure there's an issue somewhere that tracks providing more details here

> when I comment out either of the two event checks, the test passes. So both events seem fine by themselves. Also the call tree suggests that all event parameters match.

Can you open an issue with steps to reproduce? This seems odd

**Comment by 0xDEnYO** [comment, marked as answer]
Hey Guys, 

I just found the issue. 
It was the order of the vm.expectEmit() calls.
They have to be in the same order as the events which are being emitted (which makes sense but I didnt think of that).
Maybe makes sense to add a note about that in the docs?
